docs/guides/working-with-json/json-semi-structured.md

```diff
@@ -0,0 +1,459 @@
+      "url": "https://api.github.com/repos/Helikopter-Bojowy/Exp-na-helikopterze"
+    }
+}
+```
+
+```sql
+INSERT INTO github_tmp FORMAT JSONAsObject
+{"type":"PushEvent","actor":{"avatar_url":"https://avatars.githubusercontent.com/u/81258380?",
+"display_login":"Helikopter-Bojowy","gravatar_id":"","id":81258380,"login":"Helikopter-Bojowy",
+"url":"https://api.github.com/users/Helikopter-Bojowy"},"repo":{"id":352069365,
+"name":"Helikopter-Bojowy/Exp-na-helikopterze",
+"url":"https://api.github.com/repos/Helikopter-Bojowy/Exp-na-helikopterze"}}
+```
+
+If we inspect the schema, we can see the columns have automatically been inferred and added:
+
+```sql
+SET describe_extend_object_types=1;
+DESCRIBE github_tmp;
+
+Tuple(actor Tuple(avatar_url String, display_login String, gravatar_id String, 
+id Int32, login String, url String), repo Tuple(id Int32, name String, url String), 
+type String)
+```
+
+### Changing Columns
+
+Despite best efforts, JSON is often inconsistent in types. Whilst some data stores, such as Kafka, can enforce a schema on JSON this is often not enforced. As a result, ClickHouse can receive the same field in multiple types. This often requires unifying types. Consider the following example:
+
+```json
+{
+  "type": "PushEvent",
+  "actor": {
+    "id": 10
+  }
+}
+```
+
+Here `actor.id` is an integer. If inserted to a table, it will be mapped to an Int8 as shown below:
+
+```sql
+SET allow_experimental_object_type=1;
+CREATE table github_types ( event JSON ) ENGINE = MergeTree ORDER BY tuple();
+
+INSERT INTO github_types FORMAT JSONAsObject
+{"type":"PushEvent","actor":{"id":10}}
+
+SET describe_extend_object_types=1;
+DESCRIBE github_types;
+
+Tuple(actor Tuple(id Int8), type String)
+```
+
+Now Github has alot more users than can be represented by an Int8. A typical user id is much larger. Consider the more realistic example below:
+
+```sql
+INSERT INTO github_types FORMAT JSONAsObject
+{"type":"PushEvent","actor":{"id":93110249}}
+```
+
+As shown the id field is now represented as an Int32.
+
+```sql
+SET describe_extend_object_types=1;
+DESCRIBE github_types;
+
+Tuple(actor Tuple(id Int32), type String)
+```
+
+Suppose that Github decides that ids can be alphanumeric, or more realistic a value is inserted as a string e.g.
+
+```json
+{
+    "type": "PushEvent",
+    "actor": {
+      "id": "81258380"
+    }
+}
+```
+
+```sql
+INSERT INTO github_types FORMAT JSONAsObject
+{"type":"PushEvent","actor":{"id":"81258380"}}
+
+SET describe_extend_object_types=1;
+DESCRIBE github_types;
+
+Tuple(actor Tuple(id String), type String)
+```
+
+As shown, ClickHouse is now forced to represent the `actor.id` column as a string.
+
+This sort of coercion is supported for most types that have variable representation e.g. Int, Float. If necessary, ClickHouse will unify to the higher bit type that allows all current values to be represented. If necessary, converting to a `String` represents the least precise definition.
```

Comment:
Side note: is there any setting to control this behaviour ? (like force CH to trigger an error instead of unifying types)

Reply:
not currently..but i think plans to add. @alexey-milovidov is aware of the challenges of this.